object/slicer/slicer.go

```diff
@@ -0,0 +1,596 @@
+
+	_, err = writeInMemObject(x.signer, x.w, childHeader, nil, childMeta)
+	if err != nil {
+		return rootID, fmt.Errorf("write linking object: %w", err)
+	}
+
+	return rootID, nil
+}
+
+// InitPayloadStream works similar to Slice but provides PayloadWriter allowing
+// the caller to write data himself.
+func (x *Slicer) InitPayloadStream(attributes ...string) (*PayloadWriter, error) {
+	res := &PayloadWriter{
+		stream:       x.w,
+		signer:       x.signer,
+		container:    x.cnr,
+		owner:        x.owner,
+		currentEpoch: x.opts.currentNeoFSEpoch,
+		sessionToken: x.sessionToken,
+		attributes:   attributes,
+	}
+
+	res.buf.Grow(int(x.childPayloadSizeLimit()))
+	res.rootMeta.reset()
+	res.currentWriter = newLimitedWriter(io.MultiWriter(&res.buf, &res.rootMeta), x.childPayloadSizeLimit())
+
+	return res, nil
+}
+
+// PayloadWriter is a single-object payload stream provided by Slicer.
+type PayloadWriter struct {
+	stream ObjectWriter
+
+	signer       neofscrypto.Signer
+	container    cid.ID
+	owner        user.ID
+	currentEpoch uint64
+	sessionToken *session.Object
+	attributes   []string
+
+	buf bytes.Buffer
+
+	rootMeta  dynamicObjectMetadata
+	childMeta dynamicObjectMetadata
+
+	currentWriter limitedWriter
+
+	withSplit bool
+
+	writtenChildren []oid.ID
+}
+
+// Write writes next chunk of the object data. Concatenation of all chunks forms
+// the payload of the final object. When the data is over, the PayloadWriter
+// should be closed with one of the corresponding methods.
+func (x *PayloadWriter) Write(chunk []byte) (int, error) {
+	if len(chunk) == 0 {
+		// not explicitly prohibited in the io.Writer documentation
+		return 0, nil
+	}
+
+	n, err := x.currentWriter.Write(chunk)
+	if err == nil || !errors.Is(err, errOverflow) {
+		return n, err
+	}
+
+	if !x.withSplit {
+		err = x.writeIntermediateChild(x.rootMeta)
+		if err != nil {
+			return n, fmt.Errorf("write 1st child: %w", err)
+		}
+
+		x.currentWriter.reset(io.MultiWriter(&x.buf, &x.rootMeta, &x.childMeta))
+		x.withSplit = true
+	} else {
+		err = x.writeIntermediateChild(x.childMeta)
+		if err != nil {
+			return n, fmt.Errorf("write next child: %w", err)
+		}
+
+		x.currentWriter.resetProgress()
+	}
+
+	x.buf.Reset()
+	x.childMeta.reset()
+
+	n2, err := x.Write(chunk[n:]) // here n > 0 so infinite recursion shouldn't occur
+
+	return n + n2, err
+}
+
+// CloseAndRecv closes payload stream and returns oid.ID representing unique
+// identifier of the resulting object in the configured container.
+func (x *PayloadWriter) CloseAndRecv() (oid.ID, error) {
+	var res oid.ID
+	if x.withSplit {
+		return res, x.writeLastChild(x.childMeta, func(rootID oid.ID) {
+			res = rootID
+		})
+	}
+	return res, x.writeLastChild(x.rootMeta, func(rootID oid.ID) {
+		res = rootID
+	})
+}
+
+// Close is an analogue of CloseAndRecv which ignores resulting object ID. Use
+// it as io.Closer when a direct reference to the object is not required (for
+// example, with dynamic resolution via an attribute).
```

Comment:
Why don't you keep it in the `PayloadWriter` and allow to retrieve after `Close`? It's pretty natural for some content-generating app to use `WriterCloser` in a library, but outside of it you should be able to get an ID.

Reply:
ok, lets do it so